### PR TITLE
Allow ClassDB to create a Object without postinitialization for GDExtension.

### LINKED ISSUE #91023: Editor Crash when duplicating a GDExtension node if it have an internal child node event if it is added in constructor.
### Tested versions

4.3dev[7529c0bec597d70bc61975a82063bb5112ac8879]

### System information

Windows 10, Vulkan, GTX1060

### Issue description

Editor Crah when duplicating a GDExtension node(A) if it have an internal child node(B) event if it is added in constructor.
Because the flag `data.in_constructor` already be cleared when A adding B in A's constructor.

This can be fixed by [#91019](https://github.com/godotengine/godot/pull/91019) or  [#91018](https://github.com/godotengine/godot/pull/91018).

### Steps to reproduce

1. Create a GDExtension node which have an internal node, for example:
```c++
class MyNode : public Node {
	GDCLASS(MyNode, Node)

	Node *child;

protected:
	static void _bind_methods() {}

public:
	MyNode() {
		child = memnew(Node);
		add_child(child);
	}
};
```
2. Add a `MyNode` node in editor node tree. than duplicate this node.
3. The editor will crash.

-----------------------
Please download the attached file for testing, the test project is in `test_project` folder.
If you are not using windows, please compile GDExtension plugin by youself.

### Minimal reproduction project (MRP)

[bug report.zip](https://github.com/godotengine/godot/files/15066146/bug.report.zip)

### PR DESCRIPTION
**Pull request by Daylily-Zeleen**
Currently, the flow of creating GDExtension object like this:

1. Start instantiate a GDExtension object by `ObjectGDExtension::create_instance` in ClassDB.
2. In the constructor of GDExtension object base class, `Wrapped`, will create its `_owner` godot object by `ClassInfo::creation_func`.
3. In `ClassInfo::creation_func`, the owner godot object will notify `NOTIFICATION_POSTINITIALIZE` by `Node::_postinitialize` in the `memnew` in godot.
4. After construction of GDExtension object is completed, it will notify `NOTIFICATION_POSTINITIALIZE` by `Wrapped::_postinitialize` in the `memnew` in GDExtension .

This will bring two issues:
1. The GDExtension object owner already be postinitialized before it is constructed. It means that a GDExtension Node can't add an internal Node in its constructor like in godot, because it owner's flag `data.in_constructor` already be cleared.
2. The postinitialization of GDExtension object owner  will be do twice, because notify `NOTIFICATION_POSTINITIALIZE` in GDExtension is through its owner.

----------------------------------------
This pr do these things:
1. Seperate Object's initialization and postinitialization.
6. Allow ClassDB create a Object without postinitialization to be a GDExtension object owner, it will be postinitialized after constructing the GDExtension object completly.
7. Modify GodotSharp to fit above changes.

--------------------------
Fixes https://github.com/godotengine/godot/issues/91023

### CODE REVIEW

**Comment by Daylily-Zeleen** [comment]
@AThousandShips I think this pr should be add a `bug` label.



> 1. The GDExtension object owner already be postinitialized before it is constructed. It means that a GDExtension Node can't add an internal Node in its constructor like in godot, because it owner's flag `data.in_constructor` already be cleared.

Due to this reason, the edior will crash when we copying a gdextenion Node if it have a internal node (be add in constructor but not be marked as internal).

**Comment by AThousandShips** [comment]
How could the node be created in the constructor and not be marked as internal? It won't be created if duplicated in this case, that would be the bug instead, this shouldn't be needed

**Comment by Daylily-Zeleen** [comment]
> How could the node be created in the constructor and not be marked as internal? It won't be created if duplicated in this case, that would be the bug instead, this shouldn't be needed

I means that the **GDExtension node(A)** have a **node(B)** which be added in A's constructor, the B should be an internal node. But currently, the B is not mark as an internal node, because when A adding B, because the flag `data.in_constructor` of A's owner node already be cleared.

So if we copy a A in editor, the B would not be skip correctly and lead crash.

**Comment by AThousandShips** [comment]
That should be handled by the extension IMO, just like it would have to be if you do it in GDScript, in my opinion that's a limitation, a missing feature, not a bug, especially as changing this breaks code that does handle that, so would need to be adjusted for

**Comment by AThousandShips** [comment]
Regardless if this is a bug you need to open a bug report for it so we can track and discuss it

**Comment by Daylily-Zeleen** [comment]
I rename `instantiate_extension_owner` to `instantiate_without_postinitialization`, it seems that GDScript class have the same problem, may be we can user this new method to handle and sent `NOTIFICATION_POSTINITIALIZE` after `_init()` in GDScript (may be used in C#, too).

I think this can keep the consistency between multiple programming languages, and at least can help to solves the problem of editor crash caused by the duplication of custom nodes which running in the editor (a GDExtension node or GDScrip node with @tool flag) when they have internal nodes (GDExtension(c++ binding) node can be solved, GDScript node need more efforts).

**Comment by AThousandShips** [comment]
I think the other option, being less invasive, might be safer, and the extension team would be best suited to decide what milestone to direct this to, depending on the impact of the changes to initialization 

While crashes should be avoided the exact solution and cause and safe use should be discussed, to provide the best solution

File: core/extension/gdextension_interface.h
Comment (AThousandShips) [inline core/extension/gdextension_interface.h:2614]:
```suggestion
 * "NOTIFICATION_POSTINITIALIZE" must be sent after construction.
```

**Comment by Daylily-Zeleen** [comment]
Accroding to @dsnopek's [comment](https://github.com/godotengine/godot/pull/91019#discussion_r1587807046), I add `GDExtensionClassCreationInfo4` and `GDExtensionClassCreateInstance2` for gdextension.

If we can merge this pr in 4.3, I think we should merge `GDExtensionClassCreationInfo4` into `GDExtensionClassCreationInfo3`.

I would try to implement similar feature for [#91019](https://github.com/godotengine/godot/pull/91019) later.

File: core/object/class_db.cpp
Comment (dsnopek) [inline core/object/class_db.cpp:534]: Why is this doing something custom, rather than `ERR_PRINT_ONCE()`?

File: core/object/class_db.h
Comment (dsnopek) [inline core/object/class_db.h:137]:
```suggestion
		// The bool argument indicates the need to postinitialize.
```

File: core/object/object.cpp
Comment (dsnopek) [inline core/object/object.cpp:212]:
```suggestion
	_class_name_ptr = _get_class_namev(); // Set the direct pointer, which is much faster to obtain, but can only happen after initialized.
```

File: core/extension/gdextension_interface.h
Comment (dsnopek) [inline core/extension/gdextension_interface.h:2657]:
```suggestion
 * @deprecated in Godot 4.4. Use `classdb_construct_object2` instead.
```

File: core/extension/gdextension_interface.h
Comment (dsnopek) [inline core/extension/gdextension_interface.h:2671]:
```suggestion
 * @since 4.4
```

File: core/extension/gdextension_interface.h
Comment (dsnopek) [inline core/extension/gdextension_interface.h:2748]:
```suggestion
 * @deprecated in Godot 4.4. Use `classdb_register_extension_class4` instead.
```

File: core/extension/gdextension_interface.h
Comment (dsnopek) [inline core/extension/gdextension_interface.h:2763]:
```suggestion
 * @since 4.4
```

**Review by dsnopek — COMMENTED** [review]
Thanks for your continued work on this!

> If we can merge this pr in 4.3, I think we should merge GDExtensionClassCreationInfo4 into GDExtensionClassCreationInfo3.

I think we should save this for 4.4 - Godot 4.3 is basically in feature freeze at this point (with a couple of specific possible exceptions, but this PR isn't among them).

> I would try to implement similar feature for https://github.com/godotengine/godot/pull/91019 later.

At this point, I think I personally favor the approach in this PR over the other one. It would be nice to get some feedback from other GDExtension team members, though.

File: core/object/class_db.cpp
Comment (Daylily-Zeleen) [reply core/object/class_db.cpp:534]: If use `ERR_PRINT_ONCE()` In this situation, it would print once if there have multiple classes encount this error.

Comment (dsnopek) [reply core/object/class_db.cpp:534]: Ah, ok, that makes sense. However, I wonder if ensuring this is worth the extra code? Maybe we should just do plain `ERR_PRINT()` here without trying to avoid repeatedly printing this message? So:

```suggestion
			ERR_PRINT(vformat("Cannot make a placeholder instance of runtime class %s because its parent cannot be constructed.", ti->name));
```

Comment (Daylily-Zeleen) [reply core/object/class_db.cpp:534]: I‘m not sure, I just obey the previous implementation (`ERR_PRINT_ONCE()`). Maybe there has a situation that result in printing this error continuously?

Comment (dsnopek) [reply core/object/class_db.cpp:534]: I'm the one who added the `ERR_PRINT_ONCE()`. :-) I never actually saw it get printed continuously, I just figured it wasn't very useful to see it more than once, but I didn't really understand how `ERR_PRINT_ONCE()` works, and didn't realize it would prevent seeing the message for the different classes.

Personally, I think it'd be best to just switch to plain `ERR_PRINT()` without the extra code, and we can adjust it later if it turns out that this produces error spam in practice.

**Review by dsnopek — APPROVED** [review]
Thanks! This is looking great to me now :-)

**However, we should wait until the Godot 4.4 development cycle starts to merge this.**

File: core/extension/gdextension_interface.h
Comment (AThousandShips) [inline core/extension/gdextension_interface.h:2732]: This should also be updated in `classdb_register_extension_class` as well

**Review by AThousandShips — APPROVED** [review]

**Review by raulsntos — APPROVED** [review]
The changes to GodotSharp look good to me.

Does this need a follow-up in `godot-cpp`? As I understand it, this is an alternative to https://github.com/godotengine/godot-cpp/pull/1447 so should that be reverted after merging this one? Or, to put it another way, what should GDExtension language bindings do regarding the POSTINITIALIZED notification after this PR is merged?

**Comment by dsnopek** [comment]
> Does this need a follow-up in `godot-cpp`? As I understand it, this is an alternative to [godotengine/godot-cpp#1447](https://github.com/godotengine/godot-cpp/pull/1447) so should that be reverted after merging this one? Or, to put it another way, what should GDExtension language bindings do regarding the POSTINITIALIZED notification after this PR is merged?

Yep! Once a GDExtension binding switches to using `classdb_construct_object2()`, it'll be required to send `NOTIFICATION_POSTINITAILIZE` to the object after it's been constructed. We'll need to update godot-cpp to do this (which will include effectively reverting [godotengine/godot-cpp#1447](https://github.com/godotengine/godot-cpp/pull/1447), but that'll only be for Godot 4.4+. Existing GDExtensions built for Godot <=4.3 which use `classdb_construct_object()` (without the `2`) will continue working as they have

File: core/object/class_db.h
Comment (Daylily-Zeleen) [reply core/object/class_db.h]: I fix this, but the change is quit huge. This should be reviewed carefully.

File: core/crypto/crypto.h
Comment (dsnopek) [inline core/crypto/crypto.h:48]: I think the diff could be made a little smaller by giving the new argument a default:

```suggestion
	static CryptoKey *create(bool p_notify_postinitialize = true);
```

File: core/crypto/crypto.cpp
Comment (dsnopek) [inline core/crypto/crypto.cpp:194]: Then this change wouldn't be needed.

This applies to all the static `create()` methods that are called manually in some places.

File: core/object/class_db.h
Comment (dsnopek) [reply core/object/class_db.h]: Thanks! I skimmed the changes, and they look correct to me. Although, I do have an idea about how the diff could be made a little smaller, which I'll post in a moment.

File: core/crypto/crypto.h
Comment (dsnopek) [reply core/crypto/crypto.h]: Thanks! The updates look great to me :-)

**Review by dsnopek — APPROVED** [review]
Thanks! This is looking really great to me :-)

**(Reminder: Please don't merge this until the Godot 4.4 development cycle opens)**

File: core/io/http_client_tcp.h
Comment (akien-mga) [inline core/io/http_client_tcp.h:79]: There's a handful of `norify` typos.

**Comment by Daylily-Zeleen** [comment]
The ci looks like have bugs. Linux Minimal template partially use old code to compile.

**Comment by akien-mga** [comment]
Doesn't look like a CI bug yo me, but a bug in this PR. That build uses `deprecated=no` (`DISABLE_DEPRECATED`), I assume your PR is doing changes in a way that relies on deprecated code. You should be able to reproduce it with that SCons option.

**Comment by Daylily-Zeleen** [comment]
> Error: core/object/class_db.cpp:749:79: error: 'struct ObjectGDExtension' has no member named 'create_instance'; did you mean 'create_instance2'?
  749 |  return ti->creation_func == nullptr && (!ti->gdextension || ti->gdextension->create_instance == nullptr);
      |                                                                               ^~~~~~~~~~~~~~~
      |                                                                               create_instance2


No, the position and specific code of the error are not exist in this PR.

**Comment by akien-mga** [comment]
Your PR is 2000 commits behind `master`, you need to rebase.
GitHub CI tests PRs on top of `master`, which is where they will be merged. That means that while your PR branch might work for you, once merged it will break compilation as shown by CI.

**Comment by akien-mga** [comment]
Thanks!

**Comment by beicause** [comment]
Sorry to bother this closed PR, but I'm confused.

After #89442, Node doesn't use NOTIFICATION_POSTINITIALIZE anymore.

In the documentation, NOTIFICATION_POSTINITIALIZE is described as being used only internally by the engine and the script won't receive this notification. So why doesn't gdextension work the same way as scripts? Why does gdextension need NOTIFICATION_POSTINITIALIZE? Although doing postinitialization after constructor keeps consistent with the engine's internal behavior, it also relies on such internal behavior, which may change over time.

godot-rust doesn't manually emit NOTIFICATION_POSTINITIALIZE yet, so I doubt it. Please correct me if I misunderstood anything.

**Comment by dsnopek** [comment]
> Why does gdextension need NOTIFICATION_POSTINITIALIZE? Although doing postinitialization after constructor keeps consistent with the engine's internal behavior, it also relies on such internal behavior, which may change over time.

One of the goals of GDExtension is to allow extensions to do all the same stuff as internal modules. In godot-cpp, we even attempt to allow using the same APIs as used internally, so we try very hard to be consistent with the engine's internal behavior.

> godot-rust doesn't manually emit NOTIFICATION_POSTINITIALIZE yet

A GDExtension binding should only be required to manually emit it if using `classdb_register_extension_class4`. I don't know if the Rust binding has switched to using that yet or not, but when it does, it will need to start manually emitting it, otherwise any GDExtension class that extends a parent class that reacts to it may not work correctly

**Comment by Inve1951** [comment]
How do you send such a notification? I can't find an example or even just a hint in the `gdextension_interface.h` and the rust code linked above is incomprehensible (uses some `emit` function that isn't declared anywhere).

**Comment by dsnopek** [comment]
By calling [Object.notification()](https://docs.godotengine.org/en/latest/classes/class_object.html#class-object-method-notification), which is a normal method registered in `ClassDB`.

**Pull request merged**
